MoreLinq.Test/TakeLastTest.cs

```diff
@@ -57,5 +63,27 @@ public void TakeLastDisposesSequenceEnumerator()
                 seq.TakeLast(1).Consume();
             }
         }
+
+        [Test]
+        public void TakeLastOptimizedForCollections()
+        {
+            var collection = new OnceEnumerationCollection<int>(Enumerable.Range(1, 10));
+
+            foreach (var _ in collection.TakeLast(3))
+            {
+                break;
+            }
+
+            Assert.IsTrue(collection.Count > 0);
+        }
+
+        static void AssertEqual<T>(ICollection<T> input, Func<IEnumerable<T>, IEnumerable<T>> op, Action<IEnumerable<T>> action)
```

Comment:
`AssertEqual` is a misleading name for this method as it doesn't conduct an equality test itself. It leaves that up to `action` and in the case of `TakeLastWithNegativeCount`, the actual assertion is one of termination. I would suggest renaming this to `AssertTakeLast` or even `TakeLastDo` and since `op` is always going to be over `TakeLast`, change that parameter to be simply a count argument to be passed to `TakeLast`:

```c#
static void TakeLastDo<T>(ICollection<T> input, int count, Action<IEnumerable<T>> action)
{
    // Test that the behaviour does not change whether a collection
    // or a sequence is used as the source.

    action(input.TakeLast(count));
    action(input.Select(x => x).TakeLast(count));
}
```